clippy_lints/src/vec.rs

```diff
@@ -110,12 +115,9 @@ impl<'tcx> LateLintPass<'tcx> for UselessVec {
             .is_continue();
 
             if only_slice_uses {
-                self.check_vec_macro(
-                    cx,
-                    &vec_args,
-                    expr.span.ctxt().outer_expn_data().call_site,
-                    SuggestedType::Array,
-                );
+                let span = expr.span.ctxt().outer_expn_data().call_site;
+                self.spans_sure_to_lint.insert(span);
```

Comment:
Is the logic here that if all uses of the variable `let x = vec![];` adjust to a slice, then this is guaranteed to not be from a macro argument and we can always lint?

That would break down here I think:
```rs
macro_rules! m {
  ($x:expr) => {
    fn f() { let _x = $x; $x.starts_with(&[]); }
    fn f2() { let _x: Vec<i32> = $x; }
  }
}
m!(vec![1]);
```
If I'm reading this correctly, it would lint here.
All uses of `_x` in `f` adjust to a slice, *but* it's still expanded twice.

Reply:
Oh, thanks for the catch. You are correct that the logic is broken here. My intention of adding the spans_sure_to_lint set was that when sometimes we might want to lint `for _ in vec![...]`, but since there the `vec![...]` does not adjust to a slice, subseqeunt `check_expr` calls would insert None into the span_to_lint_map.

I think I will change the logic to store a set of `HirId`s that are ok to lint, since expressions after macro expansion will have distinct HirIds.